lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -73,6 +74,141 @@ static bool tryFlatteningOperands(Op op, PatternRewriter &rewriter) {
+// be aggressively stripped off.
+//
+// This function requires the narrowingCandidate to have at least 1 use.
+//
+// Returns true if IR is mutated. IR is mutated iff narrowing happens.
+static bool
+narrowOperationWidth(ExtractOp outerExtractOp, Operation *narrowingCandidate,
+                     ValueRange inputs, bool narrowTrailingBits,
+                     PatternRewriter &rewriter,
+                     std::function<Value(ArrayRef<Value>)> createOp) {
+  // If the result is never used, no point optimizing this. It will
+  // also complicated error handling in getLowestBitAndHigestBitRequired.
+  assert(!narrowingCandidate->getUsers().empty() &&
+         "narrowingCandidate must have at least one use.");
+  assert(narrowingCandidate->getNumResults() == 1 &&
+         "narrowingCandidate must have exactly one result");
+  assert(outerExtractOp &&
+         "narrowOperationWidth must be called from canonicalizing an "
+         "outerExtractOp, which serves as the root op.");
+  IntegerType narrowingCandidateType =
+      narrowingCandidate->getResultTypes().front().cast<IntegerType>();
+
+  size_t highestBitRequired;
+  size_t lowestBitRequired;
+  size_t originalOpWidth = narrowingCandidateType.getIntOrFloatBitWidth();
+
+  std::tie(lowestBitRequired, highestBitRequired) =
+      getLowestBitAndHighestBitRequired(narrowingCandidate, narrowTrailingBits,
+                                        originalOpWidth);
+
+  // Give up, because we can't make it narrower than it already is.
+  if (lowestBitRequired == 0 && highestBitRequired == originalOpWidth - 1)
+    return false;
+
+  auto loc = narrowingCandidate->getLoc();
+  size_t narrowedWidth = highestBitRequired - lowestBitRequired + 1;
+  auto narrowedType = rewriter.getIntegerType(narrowedWidth);
+  Value narrowedOperation =
+      createOp(SmallVector<Value>(llvm::map_range(inputs, [&](auto input) {
+        return rewriter.create<ExtractOp>(loc, narrowedType, input,
+                                          lowestBitRequired);
+      })));
+
+  auto getNarrowedExtractReplacement = [&](ExtractOp extractUse) {
+    auto oldLowBit = extractUse.lowBit();
+
+    assert(oldLowBit >= lowestBitRequired &&
+           "incorrectly deduced the lowest bit required in usage arguments.");
+
+    auto loc = extractUse.getLoc();
+
+    if (narrowedWidth == extractUse.getType().getWidth()) {
+      return Value(narrowedOperation);
+    }
+
+    uint32_t newLowBit = oldLowBit - lowestBitRequired;
+    Value narrowedExtract = rewriter.create<ExtractOp>(
+        loc, extractUse.getType(), narrowedOperation, newLowBit);
+    return narrowedExtract;
+  };
+
+  // Pattern rewriter requires that that the root operation (ie: outerExtractOp)
+  // in question to be updated in place, replaced or erased, so this is
+  // necessary.
+  rewriter.replaceOp(outerExtractOp,
+                     getNarrowedExtractReplacement(outerExtractOp));
+
+  // However, for other use sites, calling rewriter.replaceOp or
```

Comment:
You can use replaceOp on any op, it isn't required to be the root operation. If any bugs arise, that would be a bug in the pattern driver/documentation.

Reply:
I thought so, but I changed  `rewriter.replaceOp` to `RAUW` after running into some very occasional segfaults. The valgrind output is as follows. From what I can make of it, it looks like there are dangling references to none root ops in the Greedypatternrewriter worklist.

(This is the same valgrind output that follows is same as those in the PR description above).

```
fyq14@linux:~/dev/circt/build$ valgrind ./bin/circt-opt --canonicalize  ../test/Dialect/Comb/canonicalization.mlir 
==33936== Memcheck, a memory error detector
==33936== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==33936== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==33936== Command: ./bin/circt-opt --canonicalize ../test/Dialect/Comb/canonicalization.mlir
==33936== 
==33936== Invalid read of size 4
==33936==    at 0x63B95B0: mlir::Operation::getResults() (mlir/include/mlir/IR/Operation.h:293)
==33936==    by 0x63B8B04: mlir::Operation::getOpResults() (mlir/include/mlir/IR/Operation.h:297)
==33936==    by 0x63B8A24: mlir::Operation::use_empty() (mlir/include/mlir/IR/Operation.h:578)
==33936==    by 0x63B80C4: mlir::isOpTriviallyDead(mlir::Operation*) (mlir/lib/Interfaces/SideEffectInterfaces.cpp:33)
==33936==    by 0x5143937: (anonymous namespace)::GreedyPatternRewriteDriver::simplify(llvm::MutableArrayRef<mlir::Region>) (mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp:180)
==33936==    by 0x51434E3: mlir::applyPatternsAndFoldGreedily(llvm::MutableArrayRef<mlir::Region>, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig) (mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp:256)
==33936==    by 0x4F29D1B: (anonymous namespace)::Canonicalizer::runOnOperation() (mlir/lib/Transforms/Canonicalizer.cpp:48)
==33936==    by 0x541E557: mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (mlir/lib/Pass/Pass.cpp:385)
==33936==    by 0x541EB44: mlir::detail::OpToOpPassAdaptor::runPipeline(llvm::iterator_range<llvm::pointee_iterator<std::unique_ptr<mlir::Pass, std::default_delete<mlir::Pass> >*, mlir::Pass> >, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (mlir/lib/Pass/Pass.cpp:445)
==33936==    by 0x542022D: mlir::PassManager::runPasses(mlir::Operation*, mlir::AnalysisManager) (mlir/lib/Pass/Pass.cpp:689)
==33936==    by 0x5420117: mlir::PassManager::run(mlir::Operation*) (mlir/lib/Pass/Pass.cpp:669)
==33936==    by 0x4860F88: performActions(llvm::raw_ostream&, bool, bool, llvm::SourceMgr&, mlir::MLIRContext*, mlir::PassPipelineCLParser const&) (mlir/lib/Support/MlirOptMain.cpp:84)
==33936==  Address 0x9fb7aa4 is 52 bytes inside a block of size 128 free'd
==33936==    at 0x483CA3F: free (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==33936==    by 0x674C557: mlir::Operation::destroy() (mlir/lib/IR/Operation.cpp:224)
==33936==    by 0x674D114: llvm::ilist_traits<mlir::Operation>::deleteNode(mlir::Operation*) (mlir/lib/IR/Operation.cpp:432)
==33936==    by 0x66273E4: llvm::iplist_impl<llvm::simple_ilist<mlir::Operation>, llvm::ilist_traits<mlir::Operation> >::erase(llvm::ilist_iterator<llvm::ilist_detail::node_options<mlir::Operation, true, false, void>, false, false>) (llvm/include/llvm/ADT/ilist.h:268)
==33936==    by 0x6756EE4: llvm::iplist_impl<llvm::simple_ilist<mlir::Operation>, llvm::ilist_traits<mlir::Operation> >::erase(mlir::Operation*) (llvm/include/llvm/ADT/ilist.h:272)
==33936==    by 0x674D36F: mlir::Operation::erase() (mlir/lib/IR/Operation.cpp:481)
==33936==    by 0x676CF25: mlir::RewriterBase::replaceOp(mlir::Operation*, mlir::ValueRange) (mlir/lib/IR/PatternMatch.cpp:230)
==33936==    by 0x758E5D: bool narrowOperationWidth<circt::comb::AddOp, bool narrowOperationWidth<circt::comb::AddOp>(circt::comb::AddOp, mlir::ValueRange, bool, mlir::PatternRewriter&)::{lambda(llvm::ArrayRef<mlir::Value>)#1}>(circt::comb::AddOp, mlir::ValueRange, bool, mlir::PatternRewriter&, bool narrowOperationWidth<circt::comb::AddOp>(circt::comb::AddOp, mlir::ValueRange, bool, mlir::PatternRewriter&)::{lambda(llvm::ArrayRef<mlir::Value>)#1}) (CombFolds.cpp:162)
==33936==    by 0x757AB3: bool narrowOperationWidth<circt::comb::AddOp>(circt::comb::AddOp, mlir::ValueRange, bool, mlir::PatternRewriter&) (CombFolds.cpp:184)
==33936==    by 0x755143: _ZZL18narrowExtractWidthN5circt4comb9ExtractOpEN4mlir15PatternRewriterEENKUlT_E_clINS0_5AddOpEEEDaS4_ (CombFolds.cpp:356)
==33936==    by 0x7551D5: llvm::TypeSwitch<mlir::Operation*, bool>& llvm::TypeSwitch<mlir::Operation*, bool>::Case<circt::comb::AddOp, narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#1}&>(narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#1}&) (TypeSwitch.h:116)
==33936==    by 0x7522AC: llvm::TypeSwitch<mlir::Operation*, bool>& llvm::detail::TypeSwitchBase<llvm::TypeSwitch<mlir::Operation*, bool>, mlir::Operation*>::Case<circt::comb::AddOp, circt::comb::MulOp, , narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#1}>(narrowExtractWidth(circt::comb::ExtractOp, mlir::PatternRewriter)::{lambda(auto:1)#1}&&) (TypeSwitch.h:41)
==33936==  Block was alloc'd at
==33936==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==33936==    by 0x674B8E6: mlir::Operation::create(mlir::Location, mlir::OperationName, mlir::TypeRange, mlir::ValueRange, mlir::DictionaryAttr, mlir::BlockRange, unsigned int) (mlir/lib/IR/Operation.cpp:136)
==33936==    by 0x674BE56: mlir::Operation::create(mlir::Location, mlir::OperationName, mlir::TypeRange, mlir::ValueRange, mlir::DictionaryAttr, mlir::BlockRange, mlir::RegionRange) (mlir/lib/IR/Operation.cpp:94)
==33936==    by 0x674BD63: mlir::Operation::create(mlir::OperationState const&) (mlir/lib/IR/Operation.cpp:83)
==33936==    by 0x662F093: mlir::OpBuilder::createOperation(mlir::OperationState const&) (mlir/lib/IR/Builders.cpp:400)
==33936==    by 0x78F0589: (anonymous namespace)::OperationParser::parseCustomOperation(llvm::ArrayRef<std::tuple<llvm::StringRef, unsigned int, llvm::SMLoc> >) (mlir/lib/Parser/Parser.cpp:1922)
==33936==    by 0x78EC806: (anonymous namespace)::OperationParser::parseOperation() (mlir/lib/Parser/Parser.cpp:801)
==33936==    by 0x78F9111: (anonymous namespace)::OperationParser::parseBlockBody(mlir::Block*) (mlir/lib/Parser/Parser.cpp:2131)
==33936==    by 0x78F8797: (anonymous namespace)::OperationParser::parseBlock(mlir::Block*&) (mlir/lib/Parser/Parser.cpp:2099)
==33936==    by 0x78F830B: (anonymous namespace)::OperationParser::parseRegionBody(mlir::Region&, llvm::SMLoc, llvm::ArrayRef<std::pair<(anonymous namespace)::OperationParser::SSAUseInfo, mlir::Type> >, bool) (mlir/lib/Parser/Parser.cpp:2057)
==33936==    by 0x78F763E: (anonymous namespace)::OperationParser::parseRegion(mlir::Region&, llvm::ArrayRef<std::pair<(anonymous namespace)::OperationParser::SSAUseInfo, mlir::Type> >, bool) (mlir/lib/Parser/Parser.cpp:1996)
==33936==    by 0x78F46DB: (anonymous namespace)::CustomOpAsmParser::parseRegion(mlir::Region&, llvm::ArrayRef<mlir::OpAsmParser::OperandType>, llvm::ArrayRef<mlir::Type>, bool) (mlir/lib/Parser/Parser.cpp:1645)
==33936== 

```

Reply:
Can you file a bug with a repro and assign it to me? This should be made to work.